### PR TITLE
compostional ECL style output

### PR DESCRIPTION
**Pull request by GitPaean**
It is mostly for the ECL style output for compositional simulation. 

although it has been refined to a big extent, it is still in the prototyping stage.  When cleaning up are done, we will discuss the approach should be used.

### CODE REVIEW

**Comment by GitPaean** [comment]
jenkins build this opm-simulators=5687 please

**Comment by GitPaean** [comment]
jenkins build this opm-simulators=5687 please

**Comment by GitPaean** [comment]
Most of the regression failures is due to the SOIL output in the UNRST file. 

SOIL should not be output for Blackoil simulation. It should be output for compositional simulation. 

The failure [mpi.co2_ptflash_ecfv](https://ci.opm-project.org/job/opm-common-PR-builder/7605/testReport/junit/(root)/mpi/co2_ptflash_ecfv/) needs extra care.

**Comment by GitPaean** [comment]
jenkins build this opm-simulators=5687 please

**Comment by GitPaean** [comment]
jenkins build this opm-simulators=5687 please

**Comment by GitPaean** [comment]
All the regression failure is related to the extra output of `SOIL` in the UNRST file. Will fix it.

**Comment by GitPaean** [comment]
jenkins build this opm-simulators=5687 please

**Comment by GitPaean** [comment]
jenkins build this opm-simulators=5687 please

**Comment by GitPaean** [comment]
The last failure now is because that SPE5CASE1.SCH used by [mpi.compareECLFiles_flow+SPE5CASE1_DYN](https://ci.opm-project.org/job/opm-common-PR-builder/7614/testReport/(root)/mpi/compareECLFiles_flow_SPE5CASE1_DYN/) is using a compositional version of RPTRST.  I guess they are ignored before. 


image

**Comment by GitPaean** [comment]
jenkins build this please

File: opm/input/eclipse/Schedule/RSTConfig.hpp [hunk 1]
```diff
@@ -204,10 +204,12 @@ class RSTConfig
     RSTConfig() = default;
     RSTConfig(const SOLUTIONSection& solution_section,
               const ParseContext& parseContext,
+              const bool compositional,
```
Comment (GitPaean) [inline opm/input/eclipse/Schedule/RSTConfig.hpp:207]: It is possible that we can make `bool compositional` a member of `RSTConfig` to save the interface changes.

**Comment by GitPaean** [comment]
jenkins build this please

File: opm/input/eclipse/Schedule/RSTConfig.cpp [hunk 2]
```diff
@@ -194,6 +194,50 @@ bool is_RPTRST_mnemonic(const std::string& kw)
...
 }
 
+bool is_RPTRST_mnemonic_compositional(const std::string& kw)
+{
+    // Every compositional keyword we want to not simply ignore when handling
+    // RPTRST.  The list is sorted, so we can use binary_search for log(n)
+    // lookup.  It is important that the list is sorted, but these are all
+    // the keywords listed in the manual and unlikely to change at all
+    static constexpr const char* valid[] = {
+        "AIM",      "ALSURF",   "ALSTML",   "AMF",      "AQSP",    "AQPH",
```
Comment (GitPaean) [inline opm/input/eclipse/Schedule/RSTConfig.cpp:204]: Not sure how we should handle `FLOCn` and `LGLCn`, where `n`  is the component number and an integer, 
and also the  `TRMFxxxx` and `xxxx` is the tracer name and string.

I copied FLORES- from the `is_RPTRST_mnemonic`  without knowing why.

**Comment by GitPaean** [comment]
For the record for later, this PR self does not create regression failures. https://ci.opm-project.org/job/opm-common-PR-builder/7636/

**Comment by GitPaean** [comment]
I am marking this PR as ready for review now.  The PR can go in by itself, while it introduces extra interfaces that only used by https://github.com/OPM/opm-simulators/pull/5687 .

**Comment by GitPaean** [comment]
jenkins build this opm-simulators=5687 please

File: opm/input/eclipse/Schedule/RSTConfig.cpp (hunk 2 above)
Comment (bska) [reply opm/input/eclipse/Schedule/RSTConfig.cpp:204]: > I copied `FLORES-` from the `is_RPTRST_mnemonic` without knowing the rationale behind the `FLORES-`.

`FLORES-` is an OPM extension introduced in #3823.

Comment (bska) [reply opm/input/eclipse/Schedule/RSTConfig.cpp:204]: > Not sure how we should handle `FLOCn` and `LGLCn`, where `n`

We may need to pull the arrays whose names include a component number into a separate list.  I haven't given much thought to this problem yet.

Comment (GitPaean) [reply opm/input/eclipse/Schedule/RSTConfig.cpp:204]: Then I think I should NOT have FLOC- , LGLC- and TRMF- in the list. And very likely `FLOCn`, `LGLCn` and `TRMFXXXX` needs to be done differently, and they can be out of this list.  Anyway, I will wait for your suggestion and just put a note here for later record.

**Comment by GitPaean** [comment]
jenkins build this opm-simulators=5687 please

File: opm/input/eclipse/Schedule/RSTConfig.cpp [hunk 3]
```diff
@@ -420,10 +464,12 @@ std::pair<
     >
 RPTRST(const Opm::DeckKeyword&  keyword,
        const Opm::ParseContext& parseContext,
-       Opm::ErrorGuard&         errors)
+       Opm::ErrorGuard&         errors,
+       const bool               compositional = false)
 {
+    const auto F = compositional ? is_RPTRST_mnemonic_compositional : is_RPTRST_mnemonic;
```
Comment (bska) [inline opm/input/eclipse/Schedule/RSTConfig.cpp:470]: Maybe we could give this a better name than `F`?  E.g.,
```C++
const auto is_mnemonic = compositional
    ? is_RPTRST_mnemonic_compositional
    : is_RPTRST_mnemonic;
```

File: opm/input/eclipse/Schedule/RSTConfig.cpp (hunk 2 above)
Comment (bska) [reply opm/input/eclipse/Schedule/RSTConfig.cpp:204]: >>> I copied `FLORES-` from the `is_RPTRST_mnemonic` without knowing the rationale behind the `FLORES-`.
>> 
>> `FLORES-` is an OPM extension introduced in #3823.
>
> Then I think I should NOT have `FLOC-`, `LGLC-`, or `TRMF-` in the list.

I think I agree.

**Review by bska — COMMENTED** [review]
Looks good for the most part, although I think I'd prefer a more specific name in one instance.

**Comment by GitPaean** [comment]
jenkins build this opm-simulators=5687 please

**Comment by GitPaean** [comment]
jenkins build this opm-simulators=5687 please

**Comment by GitPaean** [comment]
jenkins build this opm-simulators=5687 please

**Comment by GitPaean** [comment]
jenkins build this opm-simulators=5687 please

**Review by bska — APPROVED** [review]
This looks good to me now and the build check is green with no additional warnings.  I'll merge this and its downstream counterpart into the master branch.

**Pull request merged**